orm/model/kvstore/kvstore.go

```diff
@@ -0,0 +1,56 @@
+// Package kvstore defines the abstract interfaces which ORM tables and indexes
+// use for reading and writing data against a KV-store backend.
+package kvstore
+
+import (
+	dbm "github.com/tendermint/tm-db"
+)
+
+// Reader is an interface for readonly access to a kv-store.
+type Reader interface {
+	// Get fetches the value of the given key, or nil if it does not exist.
+	// CONTRACT: key, value readonly []byte
+	Get(key []byte) ([]byte, error)
+
+	// Has checks if a key exists.
+	// CONTRACT: key, value readonly []byte
+	Has(key []byte) (bool, error)
+
+	// Iterator returns an iterator over a domain of keys, in ascending order. The caller must call
+	// Close when done. End is exclusive, and start must be less than end. A nil start iterates
+	// from the first key, and a nil end iterates to the last key (inclusive). Empty keys are not
+	// valid.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	// CONTRACT: start, end readonly []byte
+	Iterator(start, end []byte) (Iterator, error)
+
+	// ReverseIterator returns an iterator over a domain of keys, in descending order. The caller
+	// must call Close when done. End is exclusive, and start must be less than end. A nil end
+	// iterates from the last key (inclusive), and a nil start iterates to the first key (inclusive).
+	// Empty keys are not valid.
+	// CONTRACT: No writes may happen within a domain while an iterator exists over it.
+	// CONTRACT: start, end readonly []byte
+	ReverseIterator(start, end []byte) (Iterator, error)
+}
+
+// Iterator aliases github.com/tendermint/tm-db.Iterator.
+type Iterator = dbm.Iterator
+
+// Writer is an interface for writing to a kv-store.
+type Writer interface {
+	Reader
+
+	// Set sets the value for the given key, replacing it if it already exists.
+	// CONTRACT: key, value readonly []byte
+	Set(key, value []byte) error
+
+	// Delete deletes the key, or does nothing if the key does not exist.
+	// CONTRACT: key readonly []byte
+	Delete(key []byte) error
+}
+
+// Store combines reader and writer.
+type Store interface {
```

Comment:
Writer is already Reader so we can condensate it to Store { Writer }; 

No strong opinion, but I'd rename pkg kvstore to kv and rename `Reader` to `ReadonlyStore` and `Writer` to `Store`.

I think it'd sound more familiar and easier to read from a consuming pkg point of view.

Ex: 

```go
func doWrite(store kv.Store) { ... }
func doRead(store kv.ReadonlyStore) { ... }
```

Reply:
done